### ISSUE TITLE
can't open Device on Nexus 5 with Lollipop 5.0

### ISSUE THREAD

**Issue by bhack** [occipital/OpenNI2#43]
Hi,
Building an android application that makes use of a Structure Sensor through openni2.
We have tested on a OnePlus One with Android 4.4.4 and works nice, but I can't open the device on a nexus 5 with Lollipop 5.0.

On the lines
Status rc = device.open(openni::ANY_DEVICE);
We obtain the following with OpenNI::getExtendedError():
DeviceOpen using default: no devices found

both phones have OTG enabled and openni::OpenNI::initialize(); return STATUS_OK
Maybe there is some incompatibility with shared libraries and Android 5.0?

/cc @mtamburrano

**Comment by olafkaehler** [comment]
I've been struggling with OpenNI on Lollipop for a while. It does work, but not out of the box, as the included libusb does definitely not run on newer Androids (tested on 4.4.2 and 5.0.1).

To find the specific issues, I first enabled logging so that I could use "adb logcat" as a sort of debugger. I managed that by hacking the function usbi_log_v, adding a call to __android_log_print() rather than fprintf(), and enabling logging in the libusb config.h.

The specific issues I had were then all in linux_usbfs.c:
- find_usbfs_path() fails. Due to SELinux "enforcing" mode in Android 5 (this did not occur in 4.4), opening the /dev filesystem is problematic. You might have to root your device and change SELinux mode to "permissive".
- find_fd_by_name() fails. Probably also due to SELinux, /proc/self/fd is not accessible to opendir(). My hacky way around this is to run "su -c 'for x in /proc/[pid]/fd/*;do echo $x; realpath $x; done'" with fpopen and get what you need by parsing the output.

You also have to be aware that OpenNI currently seems to assume, that you found and opened the device you are interested in within the Java framework. This means that you must call the appropriate UsbManager.requestPermission() or register the intent in the AndroidManifest, and most importantly that you have to open a UsbDeviceConnection with UsbManager.openDevice(). This will open a file descriptor, which is then found by find_fd_by_name() in the hacked libusb, which will allow succesful communications with the USB devices.

**Comment by bhack** [comment]
Probably [this patched version of libusb](https://github.com/Gritzman/libusb) can solve this issue.

**Comment by bhack** [comment]
See also https://github.com/libusb/libusb/issues/58

**Comment by bhack** [comment]
@ntisserand @occipital Any news? This let the device unusable on ever more diffuse lollypop.

**Comment by bhack** [comment]
@olafkaehler Seems that with rooted device and a simple su command request your second bullet hack is not needed.

The actually situation is that OpenNI2 work only on rooted device when the app request superuser rights. 
Please add official support for lollipop!

**Comment by olafkaehler** [comment]
I've not been following up on this any further and decided to stick with the root only version for now. If the Gitzman libusb promises to work for Lollipop, it might be as easy as replacing the current libusb directory in OpenNI2 with that one. On a longer term basis, and I'm not sure how far off the Gritzman libusb that is, one should probably call the native java routines from within jni to open the device and extract the file descriptor. This would be more future-proof, I'd hope...

**Comment by bhack** [comment]
I think that someone at @occipital could get in charge of this or talk with upstream libusb.

**Comment by ntisserand** [comment]
Hey guys, thanks a ton for bringing this issue to our attention!

I'm also very glad you found a workaround.

The patched libusb you're mentioning looks interesting indeed, but I can't find it referenced anywhere else than: http://stackoverflow.com/a/27505189, and I suspect that the libusb team here http://marc.info/?l=libusb-devel is not aware of it, so I'm not sure if it'd make sense to talk to them about it just now.

However, if anybody gets an OpenNI Android build to work on a non-rooted Lollipop device, using that patched libusb, we'd be more than happy to merge a pull request!

**Comment by bhack** [comment]
@ntisserand I think that is not a solution. Are you declaring that occipital doesn't want official support structure sensor on lollipop. We cannot root android os on third party when deploying structure sensor application.

**Comment by jrpowers** [comment]
Hey guys.  Re-opening this -- we didn't intend for closing it to mean we didn't want to help solve the issue.  We do.

As a step towards this, does anyone with a Lollipop device know if adding the following to AndroidManifest.xml will yield a permission request & then successful connection?  (From what we understand, adding an intent-filter to the AndroidManifest.xml for USB_DEVICE_ATTACHED should prompt the user to give your app permissions.)

We're acquiring a Lollipop device to help test too, but would be great if someone can confirm.  Thanks!

Add to your AndroidManifest.xml:

<application><activity ...>
<intent-filter>
<action android:name="android.hardware.usb.action.USB_DEVICE_ATTACHMENT" />
</intent-filter>
<meta-data a
ndroid:name="android.hardware.usb.action.USB_DEVICE_ATTACHED" android:resource="@xml/device_filter" />

Also, add this file:

res/xml/device_filter.xml, and fill it with:

<?xml version="1.0" encoding="utf-8"?>
<resources>
<usb-device vendor-id="7463" product-id="1536" />
<usb-device vendor-id="7463" product-id="1537" />
<usb-device vendor-id="7463" product-id="1538" />
<usb-device vendor-id="7463" product-id="1539" />
<usb-device vendor-id="7463" product-id="1540" />
<usb-device vendor-id="7463" product-id="1541" />
<usb-device vendor-id="7463" product-id="1542" />
</resources>

**Comment by olafkaehler** [comment]
I've tried that and it doesn't help. It does change things slightly, however, and as soon as I plug in the Structure Sensor the device now pops up a message about starting the app. Also, the UsbManager.requestPermission() does no longer pop up a message asking the user, but just returns immediately. Still, from my understanding the device file can only be opened with some special magic, which presumably happens somewhere outside the app in whichever Android kernel function deals with the requestPermission and returns the reply.

**Comment by jrpowers** [comment]
We have past experience that this XML change does indeed work on pre-Lollipop.  By chance, do you know if this worked in your setup pre-Lollipop and no longer works now?

**Comment by olafkaehler** [comment]
I think I had tried the intent-filter on 4.4.2 with the same effect - it didn't work.

But even if it had, the libusb issues I mentioned earlier are definitely genuine problems and quite independent from opening any USB devices. libusb wants to access the /dev directory, and with the SELinux setting "enforcing" it's not allowed to do that anymore. "Enforcing" is the default on Lollipop. Also, I've not quite worked out how and why, but the "find_file_by_name" function definitely fails to open the /proc/[pid]/fd/ directory, presumably also due to stricter SELinux settings. This happened to me in both 4.4.2 and 5.0.

So the actual problem is getting the device fd from the java UsbManager and related stuff into the jni and libusb. The previous method with the find_file_by_name function would have been great, but is broken now. Either the fd has to be passed in some other way (which may break again in future Android versions), or the Android implementation of libusb should call the "native" Android Java functions that are guaranteed to work for opening the devices.

**Comment by bhack** [comment]
Yes the real issue is what @olafkaehler described on new selinux/android enforcing. Libusb cannot access so indiscriminately to system dir in his relative method to find the file descriptor.

**Comment by bhack** [comment]
@jrpowers Have you acquired the lollipop device? I think it is almost in every shop now.

**Comment by jrpowers** [comment]
Yes, we've purchased one and have gotten streaming working on it (thanks to some help from Lynx Labs on that front, by the way), and are trying to think of the best strategy to update OpenNI.

**Comment by bhack** [comment]
@jrpowers Are you working with nvidia shield like Lynx Labs?

**Comment by jrpowers** [comment]
They're helping us investigate that device while we investigate on other devices in parallel.  The exact device we have purchased for this work lately is the Nexus 9.  Lots of work is ongoing.

**Comment by kuldeepdhaka** [comment]
here is my version of libusb modified with proper support.
https://github.com/kuldeepdhaka/libusb/tree/android-open2

**Comment by jselikof** [comment]
Thanks for sharing your efforts, and for clearly documenting your approach, we'll definitely be looking at the details!

**Comment by olafkaehler** [comment]
Right... "Lots of work is ongoing", but so far nothing has found its way to the github repository. So what are the chances of getting support for Android Lollipop without root access?

**Comment by bhack** [comment]
Yes more and more tablets and smartphones are upgrading OTA to Lollipop. So @jrpowers  and @jselikof can we have an @occipital feedback?

**Comment by onauparc** [comment]
:+1:

**Comment by lenlen** [comment]
:+1:

**Comment by deli4iled** [comment]
:+1:

**Comment by trotlinebeercan** [comment]
Dustin with Lynx here.

We're currently working with Occipital to bring in the changes we had to make to both libusb and OpenNI to get the Structure working on Lollipop. You can pull and view the changes at https://github.com/trotlinebeercan/OpenNI2. (run ReleaseVersion.py android in Packaging)

The main change made was in relation to how the driver libraries are built and used. We've seen errors from OpenNI reading the binary files from SELinux (if you turn on verbose logging in OpenNI) before device enumeration even begins. Our fix was to build each driver statically into libOpenNI2.so and manually attach the symbols for the driver access functions when OpenNI is initialized over reading them from the binary file. We'd prefer to keep the libraries shared, but further testing is required to ensure this is the best and only option.

libusb was patched to have find_usbfs_path always return "/dev/bus/usb". @kuldeepdhaka has a more elegant solution.

The changes on my fork of Github should get OpenNI + Structure working on Lollipop without root, but isn't as up to spec as it should be to merge into the Occipital Github at this time. A best practices guide for using OpenNI with Lollipop and documentation on the changes + building OpenNI for Android are also on the list of action items.

**Comment by jas99** [comment]
@trotlinebeercan Nice!!

**Comment by olafkaehler** [comment]
Right, I can't see how the proposed patches so far address one of my main problems mentioned before: the "find_file_by_name" function somewhere deep down in the patched libusb definitely fails to open the /proc/[pid]/fd/ directory on the tablet I have. And after looking into this once again for a while, I found that not only someone else has the same problem, but even a solution. So here is my current best bet, working on Android 5.0.1 with SELinux mode set to "Enforcing"...

- Download the bog standard OpenNI2 repository as of today, branch master
- Get the patched version of libusb-1.0.19 from here: https://github.com/leapmotion/libusb.git branch leap-2.3.x
- Copy or symbolically link libusb-leapmotion/android/jni to OpenNI2/ThirdParty/PSCommon/XNLib/ThirdParty/libusb-leap
- "Deactivate" the default libusb-1.0.9-android by removing the Android.mk from that directory
- Make some minor adjustments to leap motion's libusb: deactivate tests and examples in Android.mk, change LOCAL_MODULE to libusb in libusb.mk
- Fix the annoying compile error in ThirdParty/PSCommon/XnLib/Source/XnLogAndroidWriter.cpp:61 (add "%s" as third argument to __android_log_print), unless you are stuck with some old version of the ndk
- Go to OpenNI2/Packaging and run "ReleaseVersion.py android"
- You should now get a working android build of OpenNI2

To use the above build, you have to:
- Make sure the application changes directory into getApplicationInfo().nativeLibraryDir, so that the individual .so driver files will be found (unless you link statically, as suggested above)
- Ask the android UsbManager for permission (i.e. requestPermission) and wait until this permission is granted (i.e. the whole IntentFilter shenanigans). This will open a file, which is eventually picked up by the patched libusb

The basic problem so far is still that the device detection has to be done twice, once in Android to ask the UsbManager for permission, and once OpenNI to find devices. This means, both of these need to know, what device IDs to look for. There would be a much better solution, as mentioned earlier in this thread and also here: http://sourceforge.net/p/libusb/mailman/message/32062916/
The idea is that libusb needs an "android backend", which does all the calls to the UsbManager by invoking Java functions from within C. I've not seen such a solution yet, have too many other things on my mind to write one myself, and last but not least I'm not paid to get some basic drivers working on phones. For now thanks to Leap Motion for coming up with a workable intermediate solution.

**Comment by jselikof** [comment]
@olafkaehler thanks for the writeup!  Which tablet(s) are you testing on?

**Comment by olafkaehler** [comment]
It's an Nvidia Shield Tablet, with standard OTA updates to 5.0.1. It's rooted now, but otherwise an off-the-shelf system.

**Comment by jselikof** [comment]
Interesting, since that is precisely what Lynx are using. You might want to look at @trotlinebeercan's fork, especially regarding the sample app.  There were some threading issues peculiar to the Shield that Dustin solved.

**Comment by touchline3d** [comment]
Hello,

We are trying to use the structure sensor on the Nvidia Shield with Lollipop 5.0. 
We can't open the device neither with the libraries and configuration files built from the trotlinebeercan's fork, nor with the ones built from the occipital fork. 

We obtain another message error, the functions
	rc = device.open(openni::ANY_DEVICE);
	openni::OpenNI::getExtendedError());
return: "Could not open "1d27/0600@2/2": Unknown Xiron Status!".

We have also tested the old OpenNI2 sources and followed this tutorial: http://www.hirotakaster.com/weblog/openframeworks-8-1-and-openni-2-2-on-android-tutorial/, which also leads to this error message.

Does someone have any idea about this problem?

**Comment by trotlinebeercan** [comment]
@touchline3d 

Are you using the [Android SDK USBManager](http://developer.android.com/reference/android/hardware/usb/UsbManager.html) API to grab permission to use the device?

Also, ensure in your source that openni::OpenNI::initialize() and all other initialization calls (i.e. creating, configuring, starting the stream) is 1: called after you ensure you have obtained permission using the above mentioned API and 2: all called on the same thread and in a sequential fashion.

**Comment by touchline3d** [comment]
@trotlinebeercan

Thank you for your quick answer!
Yes, we do call the OpenNI initialization functions after having acquired the USB access permission. We have tested the two solutions to obtain the permission: the automatic solution by using an intent filter, and the non-automatic solution by explicitly asking for permission. In each case, the function:
manager.hasPermission(device)
returns true. The problem probably lies elsewhere, any other idea?

**Comment by bhack** [comment]
Are there any news after one year?

**Comment by nh2** [comment]
I don't recall exactly whether it was relevant for this issue, but I remember that on some newer Android versin my `.so` file always ended up in some directories with alternating names. Check out this patch (hack) I wrote:

https://gist.github.com/nh2/f07ffa589868d4a321cf

I don't recall thoug if it was only relevant to Android x64 or general Android >= 5.

**Comment by jselikof** [comment]
@bhack have you tried proposed Android-L solutions above by @trotlinebeercan and @olafkaehler ?

**Comment by bhack** [comment]
@jselikof Why it is not integrated upstream?

**Comment by jselikof** [comment]
Integration will follow after sufficient testing of proposed solutions that are on the table.  It would be very helpful to hear from developers such as yourself which solution has worked, or not worked and why.  Thanks!

**Comment by bhack** [comment]
@jselikof Ok probably we will have some opportunity to test this patch in the next weeks. /cc @mtamburrano

**Comment by nh2** [comment]
FYI I always found it helpful to just modify the OpenNI2 code and log all the paths related to finding the library files - that's how I noticed that they are sometimes not in the place where OpenNI2 expected them.

**Comment by sujith123** [comment]
Hi all,
     I am Using Samsung tab3, Os is 5.0.1(Lollipop). I am using Jni libraries for accessing OpenNi, but in my cpp code I am running a code " int status = oniInitialize(ONI_API_VERSION);  ",
it is crashing the application, but it is successfully running in all android version other than Lollipop,
please any body help me to resolve this issue. I am stuck in here.

**Comment by nh2** [comment]
Was there a resolution?

**Comment by bhack** [comment]
I don't know.
